compiler/ast/checked_ast.nim

```diff
@@ -63,6 +71,12 @@ func len*(x: CheckedAst): int {.inline.} =
 func info*(x: CheckedAst): TLineInfo {.inline.} =
   x.n.info
 
+func typ*(x: CheckedAst): PType {.inline.} =
+  x.n.typ
+
+func `typ=`*(x: var CheckedAst, typ: PType) {.inline.} =
+  x.n.typ = typ
```

Comment:
There was an assertion here that disallowed setting the type of an AST that contained an error somewhere to anything else than a `tyError`, but I removed it again, as it would prevent using a non-error type for, e.g. a `(error; validExpr)` expression, which doesn't seem like a good idea.

Reply:
Hm, yeah, good call -- I hadn't considered that case. An error node should indeed not be assigned a type other than `tyError`.

I think blocking is the way to go here. Allowing one to cross the the diagnostic-to-AST border when assigning the type feels unintuitive and likely to cause issues (e.g. accidentally modifying the type of the stored `wrongNode`).